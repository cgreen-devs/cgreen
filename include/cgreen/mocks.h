#ifndef MOCKS_H
#define MOCKS_H

#include <cgreen/internal/mocks_internal.h>
#include <cgreen/internal/stringify_token.h>
#include <cgreen/constraint.h>
#include <cgreen/reporter.h>

#ifdef __cplusplus
namespace cgreen {
    extern "C" {
#endif

/* Program the mocked functions to expect (or not) calls when parameters match constraints...

   expect(<function>, when(<parameter>, <constraint>), ...

   expect(<function>, when(<parameter>, <constraint>), will_return(<value>));
*/
#define expect(f, ...) expect_(get_test_reporter(), STRINGIFY_TOKEN(f), __FILE__, __LINE__, \
                               (Constraint *)__VA_ARGS__ +0, (Constraint *)0)
#define always_expect(f, ...) always_expect_(get_test_reporter(), STRINGIFY_TOKEN(f), __FILE__, __LINE__, \
                                             (Constraint *)__VA_ARGS__ +0, (Constraint *)0)
#define never_expect(f, ...) never_expect_(get_test_reporter(), STRINGIFY_TOKEN(f), __FILE__, __LINE__, \
                                           (Constraint *)__VA_ARGS__ +0, (Constraint *)0)


#ifdef _MSC_VER
// another workaround for fundamental variadic macro deficiencies in Visual C++ 2012
#define mock(...) PP_NARG(__VA_ARGS__)(get_test_reporter(), __func__, __FILE__, __LINE__, #__VA_ARGS__ "", \
                                       __VA_ARGS__)
#else
#define mock(...) PP_NARG(__VA_ARGS__)(get_test_reporter(), __func__, __FILE__, __LINE__, #__VA_ARGS__ "", \
                                       __VA_ARGS__ +0)
#endif

#define when(parameter, constraint) when_(#parameter, constraint)

#define times(number_times_called) times_(number_times_called)

/* Make Cgreen mocks strict, loose or learning */
typedef enum { strict_mocks = 0, loose_mocks = 1, learning_mocks = 2 } CgreenMockMode;
extern void cgreen_mocks_are(CgreenMockMode mode);

extern const int UNLIMITED_TIME_TO_LIVE;

#ifdef __cplusplus
    }
}
#endif

#endif
