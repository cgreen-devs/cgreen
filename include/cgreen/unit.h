#ifndef UNIT_HEADER
#define UNIT_HEADER

#include "internal/unit_implementation.h"

#ifdef __cplusplus
namespace cgreen {
    extern "C" {
#endif

/* BDD style: Describe the Subject Under Test, or context, by name */
#define Describe(subject) DescribeImplementation(subject)

/* BDD style: Run this before any tests for that SUT or in that context */
#define BeforeEach(subject) BeforeEachImplementation(subject)

/* BDD style: Run this after any tests for that SUT or in that context */
#define AfterEach(subject) AfterEachImplementation(subject)

/* NOTE if you use BDD style all three of the above are required */
/* Then you must also use the BDD style Ensure(subject, test) */

/* TDD Style: Ensure(testname) {implementation} */
/* BDD Style: Ensure(subject, testname) {implementation} */
#define Ensure(...) Ensure_NARG(__VA_ARGS__)(__VA_ARGS__)

#ifdef __cplusplus
    }
}
#endif

#endif
