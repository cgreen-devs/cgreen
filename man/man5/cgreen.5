.mso www.tmac
.TH CGREEN 5
.SH NAME
cgreen \- a framework for C and C++ unit tests and mocking
.SH DESCRIPTION
.B Cgreen
is a framework for creating and running compact and easy-to-read
unittests for C and/or C++.

Here are some of its features:

.IP \[bu] 2
Fluent API resulting in very readable tests
.IP \[bu]
Expressive and clear output using the default reporter
.IP \[bu]
Fully functional mocks, both strict and loose
.IP \[bu]
Each test runs in its own process for test suite robustness
.IP \[bu]
Automatic discovery and running of tests using dynamic library inspection
.IP \[bu]
Extensive and expressive constraints for many datatypes
.IP \[bu]
BDD-flavoured test declarations with Before and After declarations
.IP \[bu]
Extensible reporting mechanism
.IP \[bu]
Fully composable test suites
.IP \[bu]
An isolated test can be run in a single process for debugging


.SH "SEE ALSO"
cgreen-runner(1)
.PP
The full documentation for
.B cgreen-runner
and
.B the Cgreen framework
is in the
.B Cgreen
manual available at
.URL https://github.com/cgreen-devs/cgreen GitHub .
