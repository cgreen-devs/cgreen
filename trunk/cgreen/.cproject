<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<?fileVersion 5.0.0?>

<cproject storage_type_id="org.eclipse.cdt.core.XmlProjectDescriptionStorage2">
	<storageModule moduleId="org.eclipse.cdt.core.settings">
		<cconfiguration id="cdt.managedbuild.toolchain.gnu.base.1521338569">
			<storageModule externalCElementFile="cdt.managedbuild.toolchain.gnu.base.1521338569_org.eclipse.cdt.core.settings" id="cdt.managedbuild.toolchain.gnu.base.1521338569" name="Default"/>
			<storageModule externalCElementFile="cdt.managedbuild.toolchain.gnu.base.1521338569_cdtBuildSystem" version="4.0.0"/>
			<storageModule externalCElementFile="cdt.managedbuild.toolchain.gnu.base.1521338569_org.eclipse.cdt.core.externalSettings"/>
		</cconfiguration>
	</storageModule>
	<storageModule moduleId="cdtBuildSystem" version="4.0.0">
		<project id="cgreen.null.1958735211" name="cgreen"/>
	</storageModule>
	<storageModule moduleId="org.eclipse.cdt.core.LanguageSettingsProviders"/>
	<storageModule moduleId="org.eclipse.cdt.make.core.buildtargets">
		<buildTargets>
			<target name="all" path="" targetID="org.eclipse.cdt.build.MakeTargetBuilder">
				<buildCommand>make</buildCommand>
				<buildArguments/>
				<buildTarget>all</buildTarget>
				<stopOnError>true</stopOnError>
				<useDefaultCommand>false</useDefaultCommand>
				<runAllBuilders>true</runAllBuilders>
			</target>
			<target name="clean" path="" targetID="org.eclipse.cdt.build.MakeTargetBuilder">
				<buildCommand>rm</buildCommand>
				<buildArguments>-rf build</buildArguments>
				<buildTarget/>
				<stopOnError>true</stopOnError>
				<useDefaultCommand>false</useDefaultCommand>
				<runAllBuilders>true</runAllBuilders>
			</target>
		</buildTargets>
	</storageModule>
	<storageModule moduleId="scannerConfiguration">
		<autodiscovery enabled="true" problemReportingEnabled="true" selectedProfileId=""/>
		<scannerConfigBuildInfo instanceId="cdt.managedbuild.toolchain.gnu.base.1521338569;cdt.managedbuild.toolchain.gnu.base.1521338569.1174655132;cdt.managedbuild.tool.gnu.c.compiler.base.349579178;cdt.managedbuild.tool.gnu.c.compiler.input.1942191604">
			<autodiscovery enabled="false" problemReportingEnabled="true" selectedProfileId="org.eclipse.cdt.make.core.GCCStandardMakePerFileProfile"/>
		</scannerConfigBuildInfo>
		<scannerConfigBuildInfo instanceId="cdt.managedbuild.toolchain.gnu.base.1521338569;cdt.managedbuild.toolchain.gnu.base.1521338569.1174655132;cdt.managedbuild.tool.gnu.cpp.compiler.base.1611546606;cdt.managedbuild.tool.gnu.cpp.compiler.input.560078280">
			<autodiscovery enabled="true" problemReportingEnabled="true" selectedProfileId=""/>
		</scannerConfigBuildInfo>
	</storageModule>
</cproject>
