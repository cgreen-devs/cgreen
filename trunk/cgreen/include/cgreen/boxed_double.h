#ifndef BOXED_DOUBLE_HEADER
#define BOXED_DOUBLE_HEADER

#include <stdint.h>

#ifdef __cplusplus
namespace cgreen {
    extern "C" {
#endif

intptr_t box_double(double d);
double as_double(intptr_t boxed_double);
double unbox_double(intptr_t boxed_double);

typedef union {
    double value;
} BoxedDouble;

#ifdef __cplusplus
    }
}
#endif

#endif /* BOXED_DOUBLE_HEADER */
