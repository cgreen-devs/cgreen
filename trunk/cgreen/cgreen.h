#include "unit.h"
#include "text_reporter.h"
#include "assertions.h"
#include <stdlib.h>
