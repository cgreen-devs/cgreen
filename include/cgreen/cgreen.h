#include <stdlib.h>
#include <cgreen/unit.h>
#include <cgreen/suite.h>
#include <cgreen/text_reporter.h>
#include <cgreen/cdash_reporter.h>
#include <cgreen/cute_reporter.h>
#include <cgreen/assertions.h>
#include <cgreen/constraint_syntax_helpers.h>
#include <cgreen/runner.h>
#include <cgreen/boxed_double.h>
