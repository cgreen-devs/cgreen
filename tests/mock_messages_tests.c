#include <cgreen/boxed_double.h>
#include <cgreen/cgreen.h>
#include <cgreen/mocks.h>
#include <cgreen/unit.h>
#include <stdlib.h>

#ifdef __cplusplus
using namespace cgreen;
#endif

/* NOTE: These tests are not designed to pass, they are run to output
   messages which need to be compared to some golden output */

Describe(Mocks);
BeforeEach(Mocks) {}
AfterEach(Mocks) {}

static int integer_out(void) {
    return (int)mock();
}

static char *string_out(int p1) {
    return (char *)mock(p1);
}

static int sample_mock(int i, const char *s) {
    return (int)mock(i, s);
}

Ensure(Mocks, can_declare_function_never_called) {
    never_expect(sample_mock);

    sample_mock(0, "");
}

Ensure(Mocks, calls_beyond_expected_sequence_fail_when_mocks_are_strict) {
    expect(integer_out,
        will_return(1)
    );

    expect(integer_out,
        will_return(2)
    );

    expect(integer_out,
        will_return(3)
    );

    assert_that(integer_out(), is_equal_to(1));
    assert_that(integer_out(), is_equal_to(2));
    assert_that(integer_out(), is_equal_to(3));
    // expected fail
    assert_that(integer_out(), is_equal_to(3));
}

// These are tentative solutions to mocks returning doubles and
// trapping type errors when they do.

/* xEnsure(Mocks, traps_double_return_value_for_int_mock) { */
/*     expect(integer_out, will_return_double(3.1415926)); */
/*     integer_out(); */
/* } */

/* static double double_out() { */
/*     return mock_double(); */
/* } */

/* xEnsure(Mocks, traps_int_return_value_for_double_mock) { */
/*     expect(double_out, will_return_double(3.1415926)); */
/*     assert_that(double_out(), is_equal_to_double(3.1415926)); */
/* } */

Ensure(Mocks, failure_when_no_presets_for_default_strict_mock) {
    assert_that(integer_out(), is_equal_to(0));
}

Ensure(Mocks, failure_reported_when_expect_after_never_expect_for_same_function) {
    never_expect(integer_out);

    expect(integer_out);
}

Ensure(Mocks, failure_reported_when_expect_after_always_expect_for_same_function) {
    always_expect(integer_out,
        will_return(666)
    );

    expect(integer_out);
}

Ensure(Mocks, reports_multiple_always_expect) {
    always_expect(integer_out);
    always_expect(integer_out);
}

Ensure(Mocks, reports_multiple_never_expect) {
    never_expect(integer_out);
    never_expect(integer_out);
}

Ensure(Mocks, single_uncalled_expectation_fails_tally) {
    expect(string_out,
           will_return(5),
           when(i, is_equal_to(666)),
           when(s, is_equal_to_string("devil"))
           );
}

static double double_out(void) {
    return (double)mock();
}

static void double_in(double in) {
    mock(box_double(in));
}

/* TODO: this will not report anything until v2 */
/* Why? */
xEnsure(Mocks, reports_mock_cannot_return_double) {
    expect(double_out, will_return_double(4.123));
    double_out();
}

Ensure(Mocks, learning_mocks_emit_none_when_learning_no_mocks) {
    cgreen_mocks_are(learning_mocks);
}

Ensure(Mocks, learning_mocks_emit_pastable_code) {
    cgreen_mocks_are(learning_mocks);
    string_out(1);
    string_out(2);
    integer_out();
    integer_out();
    string_out(3);
    integer_out();
}

Ensure(Mocks, can_learn_double_expects) {
    cgreen_mocks_are(learning_mocks);
    double_in(3.14);
}

/* It would be very nice if learning mocks could survive unexpected
   terminations but since the child process is already gone when the
   parent process detects this we need to either start catching
   exceptions in the child or communicate the learned mock calls to
   the parent some other way.
*/
xEnsure(Mocks, learning_mocks_survive_termination) {
    int *ip = 0;
    cgreen_mocks_are(learning_mocks);
    string_out(1);
    *ip = 0;
}

static void simple_mocked_function(int first, int second) {
    mock(first, second);
}

Ensure(Mocks, constraint_number_of_calls_when_not_called_enough_times) {
    expect(simple_mocked_function, times(2));
    simple_mocked_function(1, 2);
}

Ensure(Mocks, constraint_number_of_calls_out_of_order_expectations_fail) {
    expect(simple_mocked_function, when(first, is_equal_to(1)), times(1));
    expect(simple_mocked_function, when(first, is_equal_to(2)), times(1));
    simple_mocked_function(2, 2);
    simple_mocked_function(1, 2);
}