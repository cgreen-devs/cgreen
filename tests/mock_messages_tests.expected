Running "mock_messages_tests" (15 tests)...
mock_messages_tests.c: Failure: Mocks -> calls_beyond_expected_sequence_fail_when_mocks_are_strict 
	Mocked function [integer_out] was called too many times

mock_messages_tests.c: Failure: Mocks -> calls_beyond_expected_sequence_fail_when_mocks_are_strict 
	Expected [integer_out()] to [equal] [3]
		actual value:			[0]
		expected value:			[3]

mock_messages_tests.c: Failure: Mocks -> can_declare_function_never_called 
	Mocked function [sample_mock] has an expectation that it will never be called, but it was

mock_messages_tests.c: Failure: Mocks -> constraint_number_of_calls_out_of_order_expectations_fail 
	Expected [[first] parameter in [simple_mocked_function]] to [equal] [1]
		actual value:			[2]
		expected value:			[1]

mock_messages_tests.c: Failure: Mocks -> constraint_number_of_calls_out_of_order_expectations_fail 
	Expected [[first] parameter in [simple_mocked_function]] to [equal] [2]
		actual value:			[1]
		expected value:			[2]

mock_messages_tests.c: Failure: Mocks -> constraint_number_of_calls_when_not_called_enough_times 
	Expected [simple_mocked_function] to [be called] [times]
		actual value:			[1]
		expected to have been called:	[2] times

mock_messages_tests.c: Failure: Mocks -> failure_reported_when_expect_after_always_expect_for_same_function 
	Mocked function [integer_out] already has an expectation that it will always be called a certain way; any expectations declared after an always expectation are invalid

mock_messages_tests.c: Failure: Mocks -> failure_reported_when_expect_after_never_expect_for_same_function 
	Mocked function [integer_out] already has an expectation that it will never be called; any expectations declared after a never call expectation are invalid

mock_messages_tests.c: Failure: Mocks -> failure_when_no_presets_for_default_strict_mock 
	Mocked function [integer_out] did not have an expectation that it would be called

mock_messages_tests.c: Failure: Mocks -> reports_multiple_always_expect 
	Mocked function [integer_out] already has an expectation and will always be called a certain way; any expectations declared after an always expectation are discarded

mock_messages_tests.c: Failure: Mocks -> reports_multiple_never_expect 
	Mocked function [integer_out] already has an expectation that it will never be called; declaring an expectation for a function after a never call expectation is not allowed

mock_messages_tests.c: Failure: Mocks -> single_uncalled_expectation_fails_tally 
	Expected call was not made to mocked function [string_out]

  "Mocks": 4 passes, 2 skipped, 12 failures in 0ms.
Completed "mock_messages_tests": 4 passes, 2 skipped, 12 failures in 0ms.
Mocks -> can_learn_double_expects : Learned mocks are
	expect(double_in, when(in, is_equal_to_double(3.140000)));
Mocks -> learning_mocks_emit_none_when_learning_no_mocks : Learned mocks are
	<none>
Mocks -> learning_mocks_emit_pastable_code : Learned mocks are
	expect(string_out, when(p1, is_equal_to(1)));
	expect(string_out, when(p1, is_equal_to(2)));
	expect(integer_out);
	expect(integer_out);
	expect(string_out, when(p1, is_equal_to(3)));
	expect(integer_out);
